### PR TITLE
azure pipelines container for azure agent v3

### PR DESCRIPTION
**Pull request by StephaneGerardVUB**
These two files are enough to build the Docker image. They must be in the same directory.

### CODE REVIEW

File: .github/workflows/publish-azureagent.yml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v3
+
+    - name: Build Image
+      uses: redhat-actions/buildah-build@v2.12
+      id: build-image
+      with:
+        image: azure-pipelines-azureagent
+        tags: latest ${{ github.sha }}
```
Comment (wpoely86) [inline .github/workflows/publish-azureagent.yml]: can we have a datetag too? 20230918 or something the like? Can we also tag it with the version of the azure agent?

File: azure-pipelines-azurezagent/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,33 @@
+FROM ubuntu:20.04
```
Comment (wpoely86) [inline azure-pipelines-azurezagent/Dockerfile]: Why not use the 22 version?

File: azure-pipelines-azurezagent/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+RUN curl -sL https://aka.ms/InstallAzureCLIDeb | bash
+
+# Can be 'linux-x64', 'linux-arm64', 'linux-arm', 'rhel.6-x64'.
+ENV TARGETARCH=linux-x64
+ENV RELEASE=3.226.3
+ENV TARBALL=pipelines-agent-$TARGETARCH-$RELEASE.tar.gz
+
+WORKDIR /azp
+
+ADD https://vstsagentpackage.azureedge.net/agent/$RELEASE/$TARBALL ./
```
Comment (wpoely86) [inline azure-pipelines-azurezagent/Dockerfile]: can we use https://github.com/microsoft/azure-pipelines-agent/releases instead?

File: azure-pipelines-azurezagent/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    ca-certificates \
+    curl \
+    git \
+    iputils-ping \
+    jq \
+    lsb-release \
+    software-properties-common \
+    libicu66
+
+RUN curl -sL https://aka.ms/InstallAzureCLIDeb | bash
```
Comment (wpoely86) [inline azure-pipelines-azurezagent/Dockerfile]: Do we actually need this?

File: azure-pipelines-azurezagent/files/start.sh [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+trap 'exit 0' EXIT
+trap 'exit 130' INT
+trap 'exit 143' TERM
+
+chmod +x ./run.sh
+
+# To be aware of TERM and INT signals call run.sh
+# Running it with the --once flag at the end will shut down the agent after the build is executed
+./run.sh "$@" & wait $!
```
Comment (wpoely86) [inline azure-pipelines-azurezagent/files/start.sh]: I see it also has `run-docker.sh` file? Should we use that?

File: azure-pipelines-azurezagent/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    apt-transport-https \
+    apt-utils \
+    ca-certificates \
+    curl \
+    git \
+    iputils-ping \
+    jq \
+    lsb-release \
+    software-properties-common \
+    libicu66
```
Comment (wpoely86) [inline azure-pipelines-azurezagent/Dockerfile]: I see a file `bin/installdependencies.sh` in the agent.

It seems to require more then what is listed here?

```bash
                apt-get update && apt-get install -y libkrb5-3 zlib1g debsums && (apt-get install -y liblttng-ust1 || apt-get install -y liblttng-ust0)

                apt-get install -y libssl3 || apt-get install -y libssl1.1 || apt-get install -y libssl1.0.2 || apt-get install -y libssl1.0.0 || \
                   (wget http://security.ubuntu.com/ubuntu/pool/main/o/openssl/libssl1.1_1.1.1f-1ubuntu2.17_amd64.deb \
                   && dpkg -i libssl1.1_1.1.1f-1ubuntu2.17_amd64.deb)
```
```

File: azure-pipelines-azurezagent/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+RUN curl -sL https://aka.ms/InstallAzureCLIDeb | bash
+
+# Can be 'linux-x64', 'linux-arm64', 'linux-arm', 'rhel.6-x64'.
+ENV TARGETARCH=linux-x64
+ENV RELEASE=3.226.3
```
Comment (wpoely86) [inline azure-pipelines-azurezagent/Dockerfile]: ideally we can let the github action pipeline call to github to figure out the latest version, use it and tag the container with it. Can we inject an ENV to the build like that?

```
curl -s -H "Accept: application/vnd.github.v3+json" https://api.github.com/repos/microsoft/azure-pipelines-agent/releases/latest | jq .name
```

File: azure-pipelines-azurezagent/Dockerfile (hunk 2 above)
Comment (StephaneGerardVUB) [reply azure-pipelines-azurezagent/Dockerfile]: With ubuntu:22.04, during the build:
`E: Unable to locate package libicu66`

File: azure-pipelines-azurezagent/Dockerfile (hunk 3 above)
Comment (StephaneGerardVUB) [reply azure-pipelines-azurezagent/Dockerfile]: What do you mean? If I choose a release in https://github.com/microsoft/azure-pipelines-agent/releases, the links to download the tarballs are pointing to the website https://vstsagentpackage.azureedge.net.

File: azure-pipelines-azurezagent/Dockerfile (hunk 4 above)
Comment (StephaneGerardVUB) [reply azure-pipelines-azurezagent/Dockerfile]: You seem to be right, if I comment it out, it still works.

File: azure-pipelines-azurezagent/files/start.sh (hunk 5 above)
Comment (StephaneGerardVUB) [reply azure-pipelines-azurezagent/files/start.sh]: Why?

File: azure-pipelines-azurezagent/Dockerfile (hunk 7 above)
Comment (StephaneGerardVUB) [reply azure-pipelines-azurezagent/Dockerfile]: It's more a job for an ARG directive than for an ENV one. I'll look at the Dockerfile documentation if we can go that far, it looks to me rather sophisticated in this context...

Comment (StephaneGerardVUB) [reply azure-pipelines-azurezagent/Dockerfile]: it's like reusing the output of a RUN directive...

Comment (StephaneGerardVUB) [reply azure-pipelines-azurezagent/Dockerfile]: You can inject arguments like this:
`docker build -t [image-name] --build-arg [arg-variable]=[value] .`

File: azure-pipelines-azurezagent/Dockerfile (hunk 2 above)
Comment (wpoely86) [reply azure-pipelines-azurezagent/Dockerfile]: no newer version available? libicu67?

File: azure-pipelines-azurezagent/files/start.sh (hunk 5 above)
Comment (wpoely86) [reply azure-pipelines-azurezagent/files/start.sh]: The file is also is different and as it explicitly mentions docker? Not sure if the difference is relevant?

File: azure-pipelines-azurezagent/Dockerfile (hunk 2 above)
Comment (StephaneGerardVUB) [reply azure-pipelines-azurezagent/Dockerfile]: after some trials and errors, the correct version is libicu70 under ubuntu 22.04

File: azure-pipelines-azurezagent/Dockerfile (hunk 3 above)
Comment (wpoely86) [reply azure-pipelines-azurezagent/Dockerfile]: aha ok

File: azure-pipelines-azurezagent/files/start.sh [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+print_header "Running Azure Pipelines agent..."
+
+trap 'exit 0' EXIT
+trap 'exit 130' INT
+trap 'exit 143' TERM
+
+# copy the config files and credentials in workdir
+cp /mnt/config/.agent ./
+cp /mnt/config/.credentials ./
+cp /mnt/config/.credentials_rsaparams ./
```
Comment (wpoely86) [inline azure-pipelines-azurezagent/files/start.sh]: don't we need to make sure the permissions on these are also correct? Either let quattor set them and use `cp -a` or fix it here?

Comment (StephaneGerardVUB) [reply azure-pipelines-azurezagent/files/start.sh]: I tried 'cp -ap' in the Dockerfile, but I got error messages... still have to understand why...

File: .github/workflows/publish-azureagent.yml (hunk 1 above)
Comment (StephaneGerardVUB) [reply .github/workflows/publish-azureagent.yml]: tags added

File: .github/workflows/publish-azureagent.yml [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+      run: |
+          curdate=$(date +%Y%m%d)
+          echo "builddate=$curdate" >> $GITHUB_ENV
+
+    - name: Build Image
+      uses: redhat-actions/buildah-build@v2.12
+      id: build-image
+      with:
+        image: azure-pipelines-azureagent
+        tags: latest ${{ github.sha }} ${{steps.getcurdate.outputs.builddate}}
```
Comment (wpoely86) [inline .github/workflows/publish-azureagent.yml]:
```suggestion
        tags: latest ${{ github.sha }} ${{steps.getcurdate.outputs.builddate}} ${{steps.getvstsagentlatest.outputs.vstsagentrelease}} 
```

File: .github/workflows/publish-azureagent.yml [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    runs-on: ubuntu-latest
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v3
+    
+    - name: Determine latest release of VSTS Agent
+      id: getvstsagentlatest
+      run: |
+          release=$(curl -s -H "Accept: application/vnd.github.v3+json" https://api.github.com/repos/microsoft/azure-pipelines-agent/releases/latest | jq -r .name)
+          echo "vstsagentrelease=$release" >> $GITHUB_ENV
```
Comment (wpoely86) [inline .github/workflows/publish-azureagent.yml]: can we also add a regular echo so the version is clearly visible in the output?

File: .github/workflows/test-build-azureagent.yml [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+      uses: actions/checkout@v3
+
+    - name: Build Image
+      uses: redhat-actions/buildah-build@v2.12
+      id: build-image
+      with:
+        image: azure-pipelines-azureagent
+        tags: latest ${{ github.sha }}
+        context: azure-pipelines-azureagent
+        build-args: |
```
Comment (wpoely86) [inline .github/workflows/test-build-azureagent.yml:38]: this needs to be sync with the other file.

In the future we should see to merge this two files so there is no duplication

Comment (StephaneGerardVUB) [reply .github/workflows/test-build-azureagent.yml:38]: Same tagging for the test?

File: .github/workflows/publish-azureagent.yml (hunk 10 above)
Comment (StephaneGerardVUB) [reply .github/workflows/publish-azureagent.yml]: done

File: azure-pipelines-azurezagent/Dockerfile (hunk 7 above)
Comment (StephaneGerardVUB) [reply azure-pipelines-azurezagent/Dockerfile]: fixed

File: .github/workflows/test-build-azureagent.yml (hunk 11 above)
Comment (StephaneGerardVUB) [reply .github/workflows/test-build-azureagent.yml:38]: done

**Review by wpoely86 — APPROVED** [review]

**Pull request merged**